// soql-parser-js/dist_esm/parser/lexer.js
import { createToken, Lexer } from 'chevrotain';
export var ReservedKeyword = createToken({
    name: 'KeywordReserved',
    pattern: Lexer.NA,
});
export var Keyword = createToken({
    name: 'Keyword',
    pattern: Lexer.NA,
});
export var DateFunction = createToken({
    name: 'DateFunction',
    pattern: Lexer.NA,
});
export var AggregateFunction = createToken({
    name: 'AggregateFunction',
    pattern: Lexer.NA,
});
export var LocationFunction = createToken({
    name: 'LocationFunction',
    pattern: Lexer.NA,
});
export var FieldsFunction = createToken({
    name: 'FieldsFunction',
    pattern: Lexer.NA,
});
export var FieldsFunctionParamIdentifier = createToken({
    name: 'FieldsFunctionParamIdentifier',
    pattern: Lexer.NA,
});
export var OtherFunction = createToken({
    name: 'OtherFunction',
    pattern: Lexer.NA,
});
export var DateLiteral = createToken({
    name: 'DateLiteral',
    pattern: Lexer.NA,
});
export var DateLiteralNotIdentifier = createToken({
    name: 'DateLiteralNotIdentifier',
    pattern: Lexer.NA,
});
export var DateNLiteral = createToken({
    name: 'DateNLiteral',
    pattern: Lexer.NA,
});
export var RelationalOperator = createToken({
    name: 'RelationalOperator',
    pattern: Lexer.NA,
});
export var SymbolIdentifier = createToken({
    name: 'SymbolIdentifier',
    pattern: Lexer.NA,
});
export var DateIdentifier = createToken({
    name: 'DateIdentifier',
    pattern: Lexer.NA,
});
export var NumberIdentifier = createToken({
    name: 'NumberIdentifier',
    pattern: Lexer.NA,
});
export var DecimalNumberIdentifier = createToken({
    name: 'DecimalNumberIdentifier',
    pattern: Lexer.NA,
});
export var IntegerNumberIdentifier = createToken({
    name: 'IntegerNumberIdentifier',
    pattern: Lexer.NA,
});
export var IdentifierNotKeyword = createToken({
    name: 'IdentifierNotKeyword',
    pattern: Lexer.NA,
});
export var UsingScopeEnumeration = createToken({
    name: 'UsingScopeEnumeration',
    pattern: Lexer.NA,
});
export var RParenMismatch = createToken({
    name: 'RParenMismatch',
    pattern: Lexer.NA,
});
var identifierRegex = /[a-zA-Z][a-zA-Z0-9_.]*/y;
function matchIdentifier(text, startOffset) {
    identifierRegex.lastIndex = startOffset;
    var execResult = identifierRegex.exec(text);
    if (execResult && execResult[0].includes('..')) {
        return null;
    }
    return execResult;
}
export var Identifier = createToken({
    name: 'Identifier',
    pattern: matchIdentifier,
    line_breaks: false,
    categories: [IdentifierNotKeyword],
    start_chars_hint: [
        'a',
        'b',
        'c',
        'd',
        'e',
        'f',
        'g',
        'h',
        'i',
        'j',
        'k',
        'l',
        'm',
        'n',
        'o',
        'p',
        'q',
        'r',
        's',
        't',
        'u',
        'v',
        'w',
        'x',
        'y',
        'z',
        'A',
        'B',
        'C',
        'D',
        'E',
        'F',
        'G',
        'H',
        'I',
        'J',
        'K',
        'L',
        'M',
        'N',
        'O',
        'P',
        'Q',
        'R',
        'S',
        'T',
        'U',
        'V',
        'W',
        'X',
        'Y',
        'Z',
    ],
});
export var StringIdentifier = createToken({ name: 'StringIdentifier', line_breaks: true, pattern: /(')(?:(?=(\\?))\2.)*?\1/ });
export var WhiteSpace = createToken({
    name: 'WhiteSpace',
    pattern: /\s+/,
    group: Lexer.SKIPPED,
});
export var And = createToken({ name: 'AND', pattern: /AND/i, longer_alt: Identifier, categories: [Keyword, ReservedKeyword] });
export var As = createToken({ name: 'AS', pattern: /AS/i, longer_alt: Identifier, categories: [Keyword, ReservedKeyword] });
export var Desc = createToken({ name: 'DESC', pattern: /DESC/i, longer_alt: Identifier, categories: [Keyword, ReservedKeyword] });
export var Asc = createToken({ name: 'ASC', pattern: /ASC/i, longer_alt: Identifier, categories: [Keyword, ReservedKeyword] });
export var OrderBy = createToken({ name: 'ORDER_BY', pattern: /ORDER BY/i, longer_alt: Identifier });
export var Cube = createToken({ name: 'CUBE', pattern: /CUBE/i, longer_alt: Identifier, categories: [Keyword, ReservedKeyword] });
export var Else = createToken({ name: 'ELSE', pattern: /ELSE/i, longer_alt: Identifier, categories: [Keyword, ReservedKeyword] });
export var Excludes = createToken({
    name: 'EXCLUDES',
    pattern: /EXCLUDES/i,
    longer_alt: Identifier,
    categories: [Keyword, ReservedKeyword],
});
export var False = createToken({ name: 'FALSE', pattern: /FALSE/i, longer_alt: Identifier, categories: [Keyword, ReservedKeyword] });
export var First = createToken({ name: 'FIRST', pattern: /FIRST/i, longer_alt: Identifier, categories: [Keyword, ReservedKeyword] });
export var From = createToken({ name: 'FROM', pattern: /FROM/i, longer_alt: Identifier, categories: [Keyword, ReservedKeyword] });
export var GroupBy = createToken({ name: 'GROUP_BY', pattern: /GROUP BY/i, longer_alt: Identifier });
export var Having = createToken({ name: 'HAVING', pattern: /HAVING/i, longer_alt: Identifier, categories: [Keyword, ReservedKeyword] });
export var In = createToken({ name: 'IN', pattern: /IN/i, longer_alt: Identifier, categories: [Keyword, ReservedKeyword] });
export var NotIn = createToken({ name: 'NOT_IN', pattern: /NOT IN/i, longer_alt: Identifier });
export var Includes = createToken({
    name: 'INCLUDES',
    pattern: /INCLUDES/i,
    longer_alt: Identifier,
    categories: [Keyword, ReservedKeyword],
});
export var Last = createToken({ name: 'LAST', pattern: /LAST/i, longer_alt: Identifier, categories: [Keyword, ReservedKeyword] });
export var Like = createToken({ name: 'LIKE', pattern: /LIKE/i, longer_alt: Identifier, categories: [Keyword, ReservedKeyword] });
export var Limit = createToken({ name: 'LIMIT', pattern: /LIMIT/i, longer_alt: Identifier, categories: [Keyword, ReservedKeyword] });
export var Not = createToken({ name: 'NOT', pattern: /NOT/i, longer_alt: Identifier, categories: [Keyword, ReservedKeyword] });
export var Null = createToken({ name: 'NULL', pattern: /NULL/i, longer_alt: Identifier, categories: [Keyword, ReservedKeyword] });
export var Nulls = createToken({ name: 'NULLS', pattern: /NULLS/i, longer_alt: Identifier, categories: [Keyword, ReservedKeyword] });
export var Or = createToken({ name: 'OR', pattern: /OR/i, longer_alt: Identifier, categories: [Keyword, ReservedKeyword] });
export var Rollup = createToken({ name: 'ROLLUP', pattern: /ROLLUP/i, longer_alt: Identifier, categories: [Keyword, ReservedKeyword] });
export var Select = createToken({ name: 'SELECT', pattern: /SELECT/i, longer_alt: Identifier, categories: [Keyword, ReservedKeyword] });
export var True = createToken({ name: 'TRUE', pattern: /TRUE/i, longer_alt: Identifier, categories: [Keyword, ReservedKeyword] });
export var Using = createToken({ name: 'USING', pattern: /USING/i, longer_alt: Identifier, categories: [Keyword, ReservedKeyword] });
export var Where = createToken({ name: 'WHERE', pattern: /WHERE/i, longer_alt: Identifier, categories: [Keyword, ReservedKeyword] });
export var With = createToken({ name: 'WITH', pattern: /WITH/i, longer_alt: Identifier, categories: [Keyword, ReservedKeyword] });
export var For = createToken({ name: 'FOR', pattern: /FOR/i, longer_alt: Identifier, categories: [Keyword, ReservedKeyword] });
export var Update = createToken({ name: 'UPDATE', pattern: /UPDATE/i, longer_alt: Identifier, categories: [Keyword, ReservedKeyword] });
export var Above = createToken({ name: 'ABOVE', pattern: /ABOVE/i, longer_alt: Identifier, categories: [Keyword, Identifier] });
export var AboveOrBelow = createToken({
    name: 'ABOVE_OR_BELOW',
    pattern: /ABOVE_OR_BELOW/i,
    longer_alt: Identifier,
    categories: [Keyword, Identifier],
});
export var At = createToken({ name: 'AT', pattern: /AT/i, longer_alt: Identifier, categories: [Keyword, Identifier] });
export var Below = createToken({ name: 'BELOW', pattern: /BELOW/i, longer_alt: Identifier, categories: [Keyword, Identifier] });
export var DataCategory = createToken({
    name: 'DATA_CATEGORY',
    pattern: /DATA CATEGORY/i,
    longer_alt: Identifier,
    categories: [Keyword],
});
export var End = createToken({ name: 'END', pattern: /END/i, longer_alt: Identifier, categories: [Keyword, Identifier] });
export var Offset = createToken({ name: 'OFFSET', pattern: /OFFSET/i, longer_alt: Identifier, categories: [Keyword, Identifier] });
export var Reference = createToken({
    name: 'REFERENCE',
    pattern: /REFERENCE/i,
    longer_alt: Identifier,
    categories: [Keyword, Identifier],
});
export var Scope = createToken({ name: 'SCOPE', pattern: /SCOPE/i, longer_alt: Identifier, categories: [Keyword, Identifier] });
export var Tracking = createToken({ name: 'TRACKING', pattern: /TRACKING/i, longer_alt: Identifier, categories: [Keyword, Identifier] });
export var Then = createToken({ name: 'THEN', pattern: /THEN/i, longer_alt: Identifier, categories: [Keyword, Identifier] });
export var Typeof = createToken({ name: 'TYPEOF', pattern: /TYPEOF/i, longer_alt: Identifier, categories: [Keyword, Identifier] });
export var View = createToken({ name: 'VIEW', pattern: /VIEW/i, longer_alt: Identifier, categories: [Keyword, Identifier] });
export var Viewstat = createToken({ name: 'VIEWSTAT', pattern: /VIEWSTAT/i, longer_alt: Identifier, categories: [Keyword, Identifier] });
export var When = createToken({ name: 'WHEN', pattern: /WHEN/i, longer_alt: Identifier, categories: [Keyword, Identifier] });
export var SecurityEnforced = createToken({
    name: 'SECURITY_ENFORCED',
    pattern: /SECURITY_ENFORCED/i,
    longer_alt: Identifier,
    categories: [Keyword, Identifier],
});
export var CalendarMonth = createToken({
    name: 'CALENDAR_MONTH',
    pattern: /CALENDAR_MONTH/i,
    longer_alt: Identifier,
    categories: [DateFunction, Identifier],
});
export var CalendarQuarter = createToken({
    name: 'CALENDAR_QUARTER',
    pattern: /CALENDAR_QUARTER/i,
    longer_alt: Identifier,
    categories: [DateFunction, Identifier],
});
export var CalendarYear = createToken({
    name: 'CALENDAR_YEAR',
    pattern: /CALENDAR_YEAR/i,
    longer_alt: Identifier,
    categories: [DateFunction, Identifier],
});
export var DayInMonth = createToken({
    name: 'DAY_IN_MONTH',
    pattern: /DAY_IN_MONTH/i,
    longer_alt: Identifier,
    categories: [DateFunction, Identifier],
});
export var DayInWeek = createToken({
    name: 'DAY_IN_WEEK',
    pattern: /DAY_IN_WEEK/i,
    longer_alt: Identifier,
    categories: [DateFunction, Identifier],
});
export var DayInYear = createToken({
    name: 'DAY_IN_YEAR',
    pattern: /DAY_IN_YEAR/i,
    longer_alt: Identifier,
    categories: [DateFunction, Identifier],
});
export var DayOnly = createToken({
    name: 'DAY_ONLY',
    pattern: /DAY_ONLY/i,
    longer_alt: Identifier,
    categories: [DateFunction, Identifier],
});
export var FiscalMonth = createToken({
    name: 'FISCAL_MONTH',
    pattern: /FISCAL_MONTH/i,
    longer_alt: Identifier,
    categories: [DateFunction, Identifier],
});
export var FiscalQuarter = createToken({
    name: 'FISCAL_QUARTER',
    pattern: /FISCAL_QUARTER/i,
    longer_alt: Identifier,
    categories: [DateFunction, Identifier],
});
export var FiscalYear = createToken({
    name: 'FISCAL_YEAR',
    pattern: /FISCAL_YEAR/i,
    longer_alt: Identifier,
    categories: [DateFunction, Identifier],
});
export var HourInDay = createToken({
    name: 'HOUR_IN_DAY',
    pattern: /HOUR_IN_DAY/i,
    longer_alt: Identifier,
    categories: [DateFunction, Identifier],
});
export var WeekInMonth = createToken({
    name: 'WEEK_IN_MONTH',
    pattern: /WEEK_IN_MONTH/i,
    longer_alt: Identifier,
    categories: [DateFunction, Identifier],
});
export var WeekInYear = createToken({
    name: 'WEEK_IN_YEAR',
    pattern: /WEEK_IN_YEAR/i,
    longer_alt: Identifier,
    categories: [DateFunction, Identifier],
});
export var Avg = createToken({ name: 'AVG', pattern: /AVG/i, longer_alt: Identifier, categories: [AggregateFunction, Identifier] });
export var Count = createToken({ name: 'COUNT', pattern: /COUNT/i, longer_alt: Identifier, categories: [AggregateFunction, Identifier] });
export var CountDistinct = createToken({
    name: 'COUNT_DISTINCT',
    pattern: /COUNT_DISTINCT/i,
    longer_alt: Identifier,
    categories: [AggregateFunction, Identifier],
});
export var Min = createToken({ name: 'MIN', pattern: /MIN/i, longer_alt: Identifier, categories: [AggregateFunction, Identifier] });
export var Max = createToken({ name: 'MAX', pattern: /MAX/i, longer_alt: Identifier, categories: [AggregateFunction, Identifier] });
export var Sum = createToken({ name: 'SUM', pattern: /SUM/i, longer_alt: Identifier, categories: [AggregateFunction, Identifier] });
export var Distance = createToken({
    name: 'DISTANCE',
    pattern: /DISTANCE/i,
    longer_alt: Identifier,
    categories: [LocationFunction, Identifier],
});
export var Geolocation = createToken({
    name: 'GEOLOCATION',
    pattern: /GEOLOCATION/i,
    longer_alt: Identifier,
    categories: [LocationFunction, Identifier],
});
export var Fields = createToken({
    name: 'FIELDS',
    pattern: /FIELDS/i,
    longer_alt: Identifier,
    categories: [FieldsFunction, Identifier],
});
export var Format = createToken({
    name: 'FORMAT',
    pattern: /FORMAT/i,
    longer_alt: Identifier,
    categories: [OtherFunction, Identifier],
});
export var Tolabel = createToken({
    name: 'toLabel',
    pattern: /TOLABEL/i,
    longer_alt: Identifier,
    categories: [OtherFunction, Identifier],
});
export var ConvertTimeZone = createToken({
    name: 'convertTimezone',
    pattern: /CONVERTTIMEZONE/i,
    longer_alt: Identifier,
    categories: [OtherFunction, Identifier],
});
export var ConvertCurrency = createToken({
    name: 'convertCurrency',
    pattern: /CONVERTCURRENCY/i,
    longer_alt: Identifier,
    categories: [OtherFunction, Identifier],
});
export var Grouping = createToken({
    name: 'GROUPING',
    pattern: /GROUPING/i,
    longer_alt: Identifier,
    categories: [OtherFunction, Identifier],
});
export var All = createToken({
    name: 'ALL',
    pattern: /ALL/i,
    longer_alt: Identifier,
    categories: [FieldsFunctionParamIdentifier, Identifier],
});
export var Custom = createToken({
    name: 'CUSTOM',
    pattern: /CUSTOM/i,
    longer_alt: Identifier,
    categories: [FieldsFunctionParamIdentifier, Identifier],
});
export var Standard = createToken({
    name: 'STANDARD',
    pattern: /STANDARD/i,
    longer_alt: Identifier,
    categories: [FieldsFunctionParamIdentifier, Identifier],
});
export var Yesterday = createToken({
    name: 'YESTERDAY',
    pattern: /YESTERDAY/i,
    longer_alt: Identifier,
    categories: [DateLiteral, Identifier, DateLiteralNotIdentifier],
});
export var Today = createToken({
    name: 'TODAY',
    pattern: /TODAY/i,
    longer_alt: Identifier,
    categories: [DateLiteral, Identifier, DateLiteralNotIdentifier],
});
export var Tomorrow = createToken({
    name: 'TOMORROW',
    pattern: /TOMORROW/i,
    longer_alt: Identifier,
    categories: [DateLiteral, Identifier, DateLiteralNotIdentifier],
});
export var LastWeek = createToken({
    name: 'LAST_WEEK',
    pattern: /LAST_WEEK/i,
    longer_alt: Identifier,
    categories: [DateLiteral, Identifier, DateLiteralNotIdentifier],
});
export var ThisWeek = createToken({
    name: 'THIS_WEEK',
    pattern: /THIS_WEEK/i,
    longer_alt: Identifier,
    categories: [DateLiteral, Identifier, DateLiteralNotIdentifier],
});
export var NextWeek = createToken({
    name: 'NEXT_WEEK',
    pattern: /NEXT_WEEK/i,
    longer_alt: Identifier,
    categories: [DateLiteral, Identifier, DateLiteralNotIdentifier],
});
export var LastMonth = createToken({
    name: 'LAST_MONTH',
    pattern: /LAST_MONTH/i,
    longer_alt: Identifier,
    categories: [DateLiteral, Identifier, DateLiteralNotIdentifier],
});
export var ThisMonth = createToken({
    name: 'THIS_MONTH',
    pattern: /THIS_MONTH/i,
    longer_alt: Identifier,
    categories: [DateLiteral, Identifier, DateLiteralNotIdentifier],
});
export var NextMonth = createToken({
    name: 'NEXT_MONTH',
    pattern: /NEXT_MONTH/i,
    longer_alt: Identifier,
    categories: [DateLiteral, Identifier, DateLiteralNotIdentifier],
});
export var Last90_days = createToken({
    name: 'LAST_90_DAYS',
    pattern: /LAST_90_DAYS/i,
    longer_alt: Identifier,
    categories: [DateLiteral, Identifier, DateLiteralNotIdentifier],
});
export var Next90_days = createToken({
    name: 'NEXT_90_DAYS',
    pattern: /NEXT_90_DAYS/i,
    longer_alt: Identifier,
    categories: [DateLiteral, Identifier, DateLiteralNotIdentifier],
});
export var ThisQuarter = createToken({
    name: 'THIS_QUARTER',
    pattern: /THIS_QUARTER/i,
    longer_alt: Identifier,
    categories: [DateLiteral, Identifier, DateLiteralNotIdentifier],
});
export var LastQuarter = createToken({
    name: 'LAST_QUARTER',
    pattern: /LAST_QUARTER/i,
    longer_alt: Identifier,
    categories: [DateLiteral, Identifier, DateLiteralNotIdentifier],
});
export var NextQuarter = createToken({
    name: 'NEXT_QUARTER',
    pattern: /NEXT_QUARTER/i,
    longer_alt: Identifier,
    categories: [DateLiteral, Identifier, DateLiteralNotIdentifier],
});
export var ThisYear = createToken({
    name: 'THIS_YEAR',
    pattern: /THIS_YEAR/i,
    longer_alt: Identifier,
    categories: [DateLiteral, Identifier, DateLiteralNotIdentifier],
});
export var LastYear = createToken({
    name: 'LAST_YEAR',
    pattern: /LAST_YEAR/i,
    longer_alt: Identifier,
    categories: [DateLiteral, Identifier, DateLiteralNotIdentifier],
});
export var NextYear = createToken({
    name: 'NEXT_YEAR',
    pattern: /NEXT_YEAR/i,
    longer_alt: Identifier,
    categories: [DateLiteral, Identifier, DateLiteralNotIdentifier],
});
export var ThisFiscalQuarter = createToken({
    name: 'THIS_FISCAL_QUARTER',
    pattern: /THIS_FISCAL_QUARTER/i,
    longer_alt: Identifier,
    categories: [DateLiteral, Identifier, DateLiteralNotIdentifier],
});
export var LastFiscalQuarter = createToken({
    name: 'LAST_FISCAL_QUARTER',
    pattern: /LAST_FISCAL_QUARTER/i,
    longer_alt: Identifier,
    categories: [DateLiteral, Identifier, DateLiteralNotIdentifier],
});
export var NextFiscalQuarter = createToken({
    name: 'NEXT_FISCAL_QUARTER',
    pattern: /NEXT_FISCAL_QUARTER/i,
    longer_alt: Identifier,
    categories: [DateLiteral, Identifier, DateLiteralNotIdentifier],
});
export var ThisFiscalYear = createToken({
    name: 'THIS_FISCAL_YEAR',
    pattern: /THIS_FISCAL_YEAR/i,
    longer_alt: Identifier,
    categories: [DateLiteral, Identifier, DateLiteralNotIdentifier],
});
export var LastFiscalYear = createToken({
    name: 'LAST_FISCAL_YEAR',
    pattern: /LAST_FISCAL_YEAR/i,
    longer_alt: Identifier,
    categories: [DateLiteral, Identifier, DateLiteralNotIdentifier],
});
export var NextFiscalYear = createToken({
    name: 'NEXT_FISCAL_YEAR',
    pattern: /NEXT_FISCAL_YEAR/i,
    longer_alt: Identifier,
    categories: [DateLiteral, Identifier, DateLiteralNotIdentifier],
});
export var NextNDays = createToken({
    name: 'NEXT_N_DAYS',
    pattern: /NEXT_N_DAYS/i,
    longer_alt: Identifier,
    categories: [DateNLiteral, Identifier],
});
export var LastNDays = createToken({
    name: 'LAST_N_DAYS',
    pattern: /LAST_N_DAYS/i,
    longer_alt: Identifier,
    categories: [DateNLiteral, Identifier],
});
export var NDaysAgo = createToken({
    name: 'N_DAYS_AGO',
    pattern: /N_DAYS_AGO/i,
    longer_alt: Identifier,
    categories: [DateNLiteral, Identifier],
});
export var NextNWeeks = createToken({
    name: 'NEXT_N_WEEKS',
    pattern: /NEXT_N_WEEKS/i,
    longer_alt: Identifier,
    categories: [DateNLiteral, Identifier],
});
export var LastNWeeks = createToken({
    name: 'LAST_N_WEEKS',
    pattern: /LAST_N_WEEKS/i,
    longer_alt: Identifier,
    categories: [DateNLiteral, Identifier],
});
export var NWeeksAgo = createToken({
    name: 'N_WEEKS_AGO',
    pattern: /N_WEEKS_AGO/i,
    longer_alt: Identifier,
    categories: [DateNLiteral, Identifier],
});
export var NextNMonths = createToken({
    name: 'NEXT_N_MONTHS',
    pattern: /NEXT_N_MONTHS/i,
    longer_alt: Identifier,
    categories: [DateNLiteral, Identifier],
});
export var LastNMonths = createToken({
    name: 'LAST_N_MONTHS',
    pattern: /LAST_N_MONTHS/i,
    longer_alt: Identifier,
    categories: [DateNLiteral, Identifier],
});
export var NMonthsAgo = createToken({
    name: 'N_MONTHS_AGO',
    pattern: /N_MONTHS_AGO/i,
    longer_alt: Identifier,
    categories: [DateNLiteral, Identifier],
});
export var NextNQuarters = createToken({
    name: 'NEXT_N_QUARTERS',
    pattern: /NEXT_N_QUARTERS/i,
    longer_alt: Identifier,
    categories: [DateNLiteral, Identifier],
});
export var LastNQuarters = createToken({
    name: 'LAST_N_QUARTERS',
    pattern: /LAST_N_QUARTERS/i,
    longer_alt: Identifier,
    categories: [DateNLiteral, Identifier],
});
export var NQuartersAgo = createToken({
    name: 'N_QUARTERS_AGO',
    pattern: /N_QUARTERS_AGO/i,
    longer_alt: Identifier,
    categories: [DateNLiteral, Identifier],
});
export var NextNYears = createToken({
    name: 'NEXT_N_YEARS',
    pattern: /NEXT_N_YEARS/i,
    longer_alt: Identifier,
    categories: [DateNLiteral, Identifier],
});
export var LastNYears = createToken({
    name: 'LAST_N_YEARS',
    pattern: /LAST_N_YEARS/i,
    longer_alt: Identifier,
    categories: [DateNLiteral, Identifier],
});
export var NYearsAgo = createToken({
    name: 'N_YEARS_AGO',
    pattern: /N_YEARS_AGO/i,
    longer_alt: Identifier,
    categories: [DateNLiteral, Identifier],
});
export var NextNFiscalQuarters = createToken({
    name: 'NEXT_N_FISCAL_QUARTERS',
    pattern: /NEXT_N_FISCAL_QUARTERS/i,
    longer_alt: Identifier,
    categories: [DateNLiteral, Identifier],
});
export var LastNFiscalQuarters = createToken({
    name: 'LAST_N_FISCAL_QUARTERS',
    pattern: /LAST_N_FISCAL_QUARTERS/i,
    longer_alt: Identifier,
    categories: [DateNLiteral, Identifier],
});
export var NFiscalQuartersAgo = createToken({
    name: 'N_FISCAL_QUARTERS_AGO',
    pattern: /N_FISCAL_QUARTERS_AGO/i,
    longer_alt: Identifier,
    categories: [DateNLiteral, Identifier],
});
export var NextNFiscalYears = createToken({
    name: 'NEXT_N_FISCAL_YEARS',
    pattern: /NEXT_N_FISCAL_YEARS/i,
    longer_alt: Identifier,
    categories: [DateNLiteral, Identifier],
});
export var LastNFiscalYears = createToken({
    name: 'LAST_N_FISCAL_YEARS',
    pattern: /LAST_N_FISCAL_YEARS/i,
    longer_alt: Identifier,
    categories: [DateNLiteral, Identifier],
});
export var NFiscalYearsAgo = createToken({
    name: 'N_FISCAL_YEARS_AGO',
    pattern: /N_FISCAL_YEARS_AGO/i,
    longer_alt: Identifier,
    categories: [DateNLiteral, Identifier],
});
export var Equal = createToken({ name: 'EQUAL', pattern: '=', categories: [RelationalOperator] });
export var NotEqual = createToken({ name: 'NOT_EQUAL', pattern: /!=|<>/, categories: [RelationalOperator] });
export var LessThan = createToken({ name: 'LESS_THAN', pattern: '<', categories: [RelationalOperator] });
export var LessThanOrEqual = createToken({ name: 'LESS_THAN_OR_EQUAL', pattern: '<=', categories: [RelationalOperator] });
export var GreaterThan = createToken({ name: 'GREATER_THAN', pattern: '>', categories: [RelationalOperator] });
export var GreaterThanOrEqual = createToken({ name: 'GREATER_THAN_OR_EQUAL', pattern: '>=', categories: [RelationalOperator] });
export var Decimal = createToken({ name: 'DECIMAL', pattern: '.', categories: [SymbolIdentifier] });
export var Colon = createToken({ name: 'COLON', pattern: ':', categories: [SymbolIdentifier] });
export var Semicolon = createToken({ name: 'SEMICOLON', pattern: ';', categories: [SymbolIdentifier] });
export var Comma = createToken({ name: 'COMMA', pattern: ',', categories: [SymbolIdentifier] });
export var Asterisk = createToken({ name: 'ASTERISK', pattern: '*', categories: [SymbolIdentifier] });
export var LParen = createToken({ name: 'L_PAREN', pattern: '(', categories: [SymbolIdentifier] });
export var RParen = createToken({ name: 'R_PAREN', pattern: ')', categories: [SymbolIdentifier] });
export var Plus = createToken({ name: 'PLUS', pattern: '+', categories: [SymbolIdentifier] });
export var Minus = createToken({ name: 'MINUS', pattern: '-', categories: [SymbolIdentifier] });
export var DateTime = createToken({
    name: 'DATETIME',
    pattern: /[0-9]{4}-[0-9]{2}-[0-9]{2}T[0-9]{2}:[0-9]{2}:[0-9]{2}(\.[0-9]+)?(\Z|\.[0-9]+Z|\+[0-9]{2}:[0-9]{2}|\-[0-9]{2}:[0-9]{2}|\+[0-9]{4}|\-[0-9]{4})/i,
    categories: [DateIdentifier],
});
export var DateToken = createToken({ name: 'DATE', pattern: /[0-9]{4}-[0-9]{2}-[0-9]{2}/, categories: [DateIdentifier] });
export var CurrencyPrefixedDecimal = createToken({
    name: 'CURRENCY_PREFIXED_DECIMAL',
    pattern: /[a-zA-Z]{3}[0-9]+\.\d+/,
    longer_alt: Identifier,
    categories: [DecimalNumberIdentifier],
});
export var SignedDecimal = createToken({
    name: 'SIGNED_DECIMAL',
    pattern: /(\-|\+)[0-9]*\.\d+/,
    categories: [NumberIdentifier, DecimalNumberIdentifier],
});
export var UnsignedDecimal = createToken({
    name: 'UNSIGNED_DECIMAL',
    pattern: /[0-9]*\.\d+/,
    categories: [NumberIdentifier, DecimalNumberIdentifier],
});
export var CurrencyPrefixedInteger = createToken({
    name: 'CURRENCY_PREFIXED_INTEGER',
    pattern: /[a-zA-Z]{3}[0-9]+/,
    longer_alt: Identifier,
    categories: [DecimalNumberIdentifier, Identifier],
});
export var SignedInteger = createToken({
    name: 'SIGNED_INTEGER',
    pattern: /(\-|\+)[0-9]+/,
    categories: [NumberIdentifier, IntegerNumberIdentifier],
});
export var GeolocationUnit = createToken({
    name: 'GEOLOCATION_UNIT',
    pattern: /'(mi|km)'/i,
    longer_alt: Identifier,
    categories: [Identifier],
});
export var UnsignedInteger = createToken({
    name: 'UNSIGNED_INTEGER',
    pattern: /0|[1-9]\d*/,
    categories: [NumberIdentifier, IntegerNumberIdentifier],
});
export var Delegated = createToken({
    name: 'Delegated',
    pattern: /DELEGATED/i,
    longer_alt: Identifier,
    categories: [UsingScopeEnumeration, Identifier],
    start_chars_hint: ['D', 'd'],
});
export var Everything = createToken({
    name: 'Everything',
    pattern: /EVERYTHING/i,
    longer_alt: Identifier,
    categories: [UsingScopeEnumeration, Identifier],
    start_chars_hint: ['E', 'e'],
});
export var MineAndMyGroups = createToken({
    name: 'MineAndMyGroups',
    pattern: /MINEANDMYGROUPS/i,
    longer_alt: Identifier,
    categories: [UsingScopeEnumeration, Identifier],
    start_chars_hint: ['M', 'm'],
});
export var Mine = createToken({
    name: 'Mine',
    pattern: /MINE/i,
    longer_alt: Identifier,
    categories: [UsingScopeEnumeration, Identifier],
    start_chars_hint: ['M', 'm'],
});
export var MyTerritory = createToken({
    name: 'MyTerritory',
    pattern: /MY_TERRITORY/i,
    longer_alt: Identifier,
    categories: [UsingScopeEnumeration, Identifier],
    start_chars_hint: ['M', 'm'],
});
export var MyTeamTerritory = createToken({
    name: 'MyTeamTerritory',
    pattern: /MY_TEAM_TERRITORY/i,
    longer_alt: Identifier,
    categories: [UsingScopeEnumeration, Identifier],
    start_chars_hint: ['M', 'm'],
});
export var Team = createToken({
    name: 'Team',
    pattern: /TEAM/i,
    longer_alt: Identifier,
    categories: [UsingScopeEnumeration, Identifier],
    start_chars_hint: ['T', 't'],
});
export var AllPrivate = createToken({
    name: 'AllPrivate',
    pattern: /ALLPRIVATE/i,
    longer_alt: Identifier,
    categories: [UsingScopeEnumeration, Identifier],
    start_chars_hint: ['A', 'a'],
});
export var allTokens = [
    WhiteSpace,
    And,
    Asc,
    As,
    OrderBy,
    Cube,
    Desc,
    Else,
    Excludes,
    False,
    First,
    From,
    Grouping,
    GroupBy,
    Having,
    Includes,
    Like,
    Limit,
    Nulls,
    Null,
    Rollup,
    Select,
    True,
    Using,
    Where,
    With,
    Update,
    Delegated,
    Everything,
    MineAndMyGroups,
    Mine,
    MyTerritory,
    MyTeamTerritory,
    Team,
    AllPrivate,
    AboveOrBelow,
    Above,
    At,
    Below,
    DataCategory,
    End,
    Offset,
    Reference,
    Scope,
    Tracking,
    Then,
    Typeof,
    Viewstat,
    View,
    When,
    SecurityEnforced,
    CalendarMonth,
    CalendarQuarter,
    CalendarYear,
    DayInMonth,
    DayInWeek,
    DayInYear,
    DayOnly,
    FiscalMonth,
    FiscalQuarter,
    FiscalYear,
    HourInDay,
    WeekInMonth,
    WeekInYear,
    Avg,
    CountDistinct,
    Count,
    Min,
    Max,
    Sum,
    Distance,
    Geolocation,
    Fields,
    Format,
    Tolabel,
    ConvertTimeZone,
    ConvertCurrency,
    Yesterday,
    Today,
    Tomorrow,
    LastWeek,
    ThisWeek,
    NextWeek,
    LastMonth,
    ThisMonth,
    NextMonth,
    Last90_days,
    Next90_days,
    ThisQuarter,
    LastQuarter,
    NextQuarter,
    ThisYear,
    LastYear,
    NextYear,
    ThisFiscalQuarter,
    LastFiscalQuarter,
    NextFiscalQuarter,
    ThisFiscalYear,
    LastFiscalYear,
    NextFiscalYear,
    NextNDays,
    LastNDays,
    NDaysAgo,
    NextNWeeks,
    LastNWeeks,
    NWeeksAgo,
    NextNMonths,
    LastNMonths,
    NMonthsAgo,
    NextNQuarters,
    LastNQuarters,
    NQuartersAgo,
    NextNYears,
    LastNYears,
    NYearsAgo,
    NextNFiscalQuarters,
    LastNFiscalQuarters,
    NFiscalQuartersAgo,
    NextNFiscalYears,
    LastNFiscalYears,
    NFiscalYearsAgo,
    GeolocationUnit,
    All,
    Custom,
    Standard,
    In,
    NotIn,
    For,
    Or,
    Last,
    Not,
    CurrencyPrefixedDecimal,
    CurrencyPrefixedInteger,
    StringIdentifier,
    Identifier,
    DateTime,
    DateToken,
    SignedDecimal,
    UnsignedDecimal,
    UnsignedInteger,
    SignedInteger,
    Equal,
    NotEqual,
    LessThanOrEqual,
    LessThan,
    GreaterThanOrEqual,
    GreaterThan,
    Decimal,
    Colon,
    Semicolon,
    Comma,
    Asterisk,
    LParen,
    RParen,
    Plus,
    Minus,
];
var SoqlLexer = new Lexer(allTokens, { ensureOptimizations: true, skipValidations: false });
export function lex(soql) {
    return SoqlLexer.tokenize(soql);
}
