// soql-parser-js/dist/src/utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getWhereValue = exports.isGroupByFn = exports.isGroupByField = exports.isOrderByFn = exports.isOrderByField = exports.isValueQueryCondition = exports.isNegationCondition = exports.isValueFunctionCondition = exports.isValueWithDateNLiteralCondition = exports.isValueWithDateLiteralCondition = exports.isValueCondition = exports.isWhereOrHavingClauseWithRightCondition = exports.isHavingClauseWithRightCondition = exports.isWhereClauseWithRightCondition = exports.isFieldSubquery = exports.isSubquery = exports.isComposeFieldTypeof = exports.isComposeFieldSubquery = exports.isComposeFieldRelationship = exports.isComposeFieldFunction = exports.isComposeField = exports.hasAlias = exports.getParams = exports.generateParens = exports.pad = exports.getAsArrayStr = exports.getLastItem = exports.getIfTrue = exports.get = exports.isNil = exports.isObject = exports.isBoolean = exports.isNumber = exports.isStringArray = exports.isString = exports.isSubqueryFromFlag = exports.isToken = void 0;
function isToken(val) {
    val = Array.isArray(val) ? val[0] : val;
    return val.image && true;
}
exports.isToken = isToken;
function isSubqueryFromFlag(val, isSubquery) {
    return isSubquery;
}
exports.isSubqueryFromFlag = isSubqueryFromFlag;
function isString(val) {
    return typeof val === 'string';
}
exports.isString = isString;
function isStringArray(val) {
    if (!val) {
        return false;
    }
    return val.every(function (item) { return isString(item); });
}
exports.isStringArray = isStringArray;
function isNumber(val) {
    return Number.isFinite(val);
}
exports.isNumber = isNumber;
function isBoolean(val) {
    return typeof val === typeof true;
}
exports.isBoolean = isBoolean;
function isObject(val) {
    return val instanceof Object;
}
exports.isObject = isObject;
function isNil(val) {
    return val === null || val === undefined;
}
exports.isNil = isNil;
function get(val, suffix, prefix) {
    return isNil(val) ? '' : "" + (prefix || '') + val + (suffix || '');
}
exports.get = get;
function getIfTrue(val, returnStr) {
    return isBoolean(val) && val ? returnStr : '';
}
exports.getIfTrue = getIfTrue;
function getLastItem(arr) {
    return arr[arr.length - 1];
}
exports.getLastItem = getLastItem;
function getAsArrayStr(val, alwaysParens) {
    if (alwaysParens === void 0) { alwaysParens = false; }
    if (Array.isArray(val)) {
        if (val.length > 0) {
            return "(" + val.join(', ') + ")";
        }
        else {
            return alwaysParens ? '()' : '';
        }
    }
    else {
        return alwaysParens ? "(" + (val || '') + ")" : val || '';
    }
}
exports.getAsArrayStr = getAsArrayStr;
function pad(val, len, left) {
    if (left === void 0) { left = 0; }
    var leftPad = left > 0 ? new Array(left).fill(' ').join('') : '';
    if (val.length > len) {
        return "" + leftPad + val;
    }
    else {
        return "" + leftPad + val + new Array(len - val.length).fill(' ').join('');
    }
}
exports.pad = pad;
function generateParens(count, character, joinCharacter) {
    if (joinCharacter === void 0) { joinCharacter = ''; }
    return isNumber(count) && count > 0 ? new Array(count).fill(character).join(joinCharacter) : '';
}
exports.generateParens = generateParens;
function getParams(functionFieldExp) {
    if (!functionFieldExp.parameters || functionFieldExp.parameters.length === 0) {
        return [];
    }
    if (isStringArray(functionFieldExp.parameters)) {
        return functionFieldExp.parameters;
    }
    if (isString(functionFieldExp.parameters[0])) {
        return [functionFieldExp.parameters[0]];
    }
    return getParams(functionFieldExp.parameters[0]);
}
exports.getParams = getParams;
function hasAlias(value) {
    return value && !isNil(value.alias);
}
exports.hasAlias = hasAlias;
function isComposeField(input) {
    return isString(input.field) && !Array.isArray(input.relationships) && !Array.isArray(input.conditions);
}
exports.isComposeField = isComposeField;
function isComposeFieldFunction(input) {
    return !isNil(input.functionName || input.fn);
}
exports.isComposeFieldFunction = isComposeFieldFunction;
function isComposeFieldRelationship(input) {
    return isString(input.field) && Array.isArray(input.relationships);
}
exports.isComposeFieldRelationship = isComposeFieldRelationship;
function isComposeFieldSubquery(input) {
    return !isNil(input.subquery);
}
exports.isComposeFieldSubquery = isComposeFieldSubquery;
function isComposeFieldTypeof(input) {
    return isString(input.field) && Array.isArray(input.conditions);
}
exports.isComposeFieldTypeof = isComposeFieldTypeof;
function isSubquery(query) {
    return isString(query.relationshipName);
}
exports.isSubquery = isSubquery;
function isFieldSubquery(value) {
    return value && value.type && value.type === 'FieldSubquery';
}
exports.isFieldSubquery = isFieldSubquery;
function isWhereClauseWithRightCondition(value) {
    return value && value.operator && value.right;
}
exports.isWhereClauseWithRightCondition = isWhereClauseWithRightCondition;
function isHavingClauseWithRightCondition(value) {
    return value && value.operator && value.right;
}
exports.isHavingClauseWithRightCondition = isHavingClauseWithRightCondition;
function isWhereOrHavingClauseWithRightCondition(value) {
    return value && value.operator && value.right;
}
exports.isWhereOrHavingClauseWithRightCondition = isWhereOrHavingClauseWithRightCondition;
function isValueCondition(value) {
    return value && isString(value.field) && isString(value.operator) && !isNil(value.value);
}
exports.isValueCondition = isValueCondition;
function isValueWithDateLiteralCondition(value) {
    return (value &&
        isString(value.field) &&
        isString(value.operator) &&
        !isNil(value.value) &&
        (value.literalType === 'DATE_LITERAL' || (Array.isArray(value.literalType) && value.literalType[0] === 'DATE_LITERAL')));
}
exports.isValueWithDateLiteralCondition = isValueWithDateLiteralCondition;
function isValueWithDateNLiteralCondition(value) {
    return value && isString(value.field) && isString(value.operator) && !isNil(value.value) && !isNil(value.dateLiteralVariable);
}
exports.isValueWithDateNLiteralCondition = isValueWithDateNLiteralCondition;
function isValueFunctionCondition(value) {
    return value && !isNil(value.fn) && isString(value.operator) && !isNil(value.value);
}
exports.isValueFunctionCondition = isValueFunctionCondition;
function isNegationCondition(value) {
    return value && isNumber(value.openParen) && isNil(value.operator) && isNil(value.field) && isNil(value.fn) && isNil(value.closeParen);
}
exports.isNegationCondition = isNegationCondition;
function isValueQueryCondition(value) {
    return value && isString(value.field) && isString(value.operator) && !isNil(value.valueQuery) && isNil(value.value);
}
exports.isValueQueryCondition = isValueQueryCondition;
function isOrderByField(value) {
    return value && !isNil(value.field);
}
exports.isOrderByField = isOrderByField;
function isOrderByFn(value) {
    return value && !isNil(value.fn);
}
exports.isOrderByFn = isOrderByFn;
function isGroupByField(value) {
    return value && !isNil(value.field);
}
exports.isGroupByField = isGroupByField;
function isGroupByFn(value) {
    return value && !isNil(value.fn);
}
exports.isGroupByFn = isGroupByFn;
function getWhereValue(value, literalType) {
    if (isNil(literalType)) {
        return value;
    }
    if (Array.isArray(literalType) && Array.isArray(value)) {
        return value.map(function (val, i) {
            return whereValueHelper(val, literalType[i]);
        });
    }
    else {
        if (Array.isArray(literalType)) {
            literalType = literalType[0];
        }
        switch (literalType) {
            case 'STRING': {
                if (Array.isArray(value)) {
                    return value.map(function (val) { return (val.startsWith("'") ? val : "'" + val + "'"); });
                }
                else {
                    return value.startsWith("'") ? value : "'" + value + "'";
                }
            }
            case 'APEX_BIND_VARIABLE': {
                return ":" + value;
            }
            default: {
                return value;
            }
        }
    }
}
exports.getWhereValue = getWhereValue;
function whereValueHelper(value, literalType) {
    switch (literalType) {
        case 'STRING': {
            return value.startsWith("'") ? value : "'" + value + "'";
        }
        default: {
            return value;
        }
    }
}
